Remove the `makeXXX` functions from `Module` that create IR instructions.

lucteo opened the pull request:
Remove the `makeXXX` functions from `Module` that create IR instructions.

Replaced them with direct constructor access.
It is envisioned that most of the IR instruction creation is done through
`Emitter`.

lucteo committed:
Remove the `makeXXX` functions from `Module` that create IR instructions.

Replaced them with direct constructor access.
It is envisioned that most of the IR instruction creation is done through
`Emitter`.

lucteo committed:
Simplify access to current `Function` in `Emitter`

lucteo commented on Sources/IR/Analysis/Module+AccessReification.swift:
```diff
@@ -131,7 +131,12 @@ extension Module {
 
-    let reified = makeAccess([k], from: s.source, correspondingTo: s.binding, in: f, at: s.site)
+    let reified = Access(
+      capabilities: [k],
```
Overall, we said that clients will use the underscore functions from `Emitter`. But, I think that we will still go back and forth a bit on this.

I would keep this PR as much as possible, to just removing the `makeXXX` functions from `Module`. But, I'm fine changing the signatures of the initializers if that makes more sense.

(as typing is not the main bottleneck, and as calling these initializers will be encapsulated anyways (hopefully, eventually), I don't have a strong preference.).

lucteo replied:
Also, the next place is one place that we are not yet using `Emitter`.
At this point, `Emitter` doesn't seem a good fit for places like this in which we just replace one instruction w/ another.

kyouko-taiga commented on Sources/IR/Emitter.swift:
```diff
@@ -89,6 +89,12 @@ struct Emitter: Sendable {
     program.ast
   }
 
+  /// The IR in which we are currently inserting.
```
Neither the comment nor the name of the property makes much sense to me. I want to call this property `insertionFunction` but I see that it clashes with another one.

In general I have come to accept that we need a lot of subscripting to interact with the various APIs of the compiler. Even in the new front-end there's a fair amount of `program.types[x]` in various places. I also believe it creates a better debugging experience because I almost never want to enter a getter when I step into an expression like `foo(bar.ham)`. Long story short my point is that I don't think avoiding subscripting is necessarily a clear cut improvement.

Looking at the rest of Emitter it seems this particular change mostly replaces `module[insertionFunction!].type(of: ...)` for `insertionIR!.type(of: ...)`. Then I would suggest to write a function `irType(of:)` and dispense with this property.

lucteo replied:
Makes sense. I can revert this change and use `irType(of:)`.

kyouko-taiga commented on Sources/IR/Emitter.swift:
```diff
@@ -3416,14 +3423,21 @@ struct Emitter: Sendable {
+      elementPath = subfield
+      address = recordAddress
     }
+    let ir = insertionIR!
```
This property is very dangerous for performance!

If Swift decides that it needs to copy `ir` because something else is being mutated in the module it may create very, very large temporaries. That's another reason why I am not a fan of removing explicit subscripting.

kyouko-taiga commented on Sources/IR/Emitter.swift:
```diff
@@ -3715,160 +3734,248 @@ extension Emitter {
   fileprivate mutating func _end_access(_ x: Operand) {
-    insert(module.makeEndAccess(x, in: insertionFunction!, at: currentSource))
+    let ir = insertionIR!
+    precondition(x.instruction.map({ ir[$0] is Access }) ?? false)
```
Another instance of the `optional.map(predicate) ?? false` that I think should be expressed more concisely.

kyouko-taiga commented on Sources/IR/Emitter.swift:
```diff
@@ -3715,160 +3734,248 @@ extension Emitter {
+      bundle: b, variants: variants,
+      parameters: t.inputs.lazy.map({ ParameterType($0.type)! }),
+      projection: RemoteType(t.output)!.bareType,
+      operands: arguments, site: currentSource))!
```
This has to go in a function.

lucteo replied:
I assume you mean everything except the call to `insert`. Is that right?

kyouko-taiga commented on Sources/IR/Operands/Instruction/AdvancedByBytes.swift:
```diff
@@ -45,20 +45,3 @@ extension AdvancedByBytes: CustomStringConvertible {
-extension Module {
-
-  /// Creates an `advanced by bytes` instruction anchored at `site` computing the `source` address
-  /// value advanced by `offset` bytes.
```
There's valuable documentation in these comments (and in those attached to other `makeXXX` functions) that should be preserved somehow.

lucteo replied:
Isn't this (advanced by `offset` bytes) documented at in the type documentation?

kyouko-taiga reviewed (changes requested):
I am not convinced this PR is an improvement.

I understand this is an intermediate change but it is loosing valuable information that I thing will be difficult to recover. At the very least it won't be easy to ensure in the next PR that the invariants currently enforced by the `makeXXX` functions as well as their documentation lands in the right place once the dust settles.

My suggestion is to move the `makeXXX` functions directly to the emitter.

lucteo commented on Sources/IR/Emitter.swift:
```diff
@@ -3416,14 +3423,21 @@ struct Emitter: Sendable {
-      return insert(s)!
+      elementPath = subfield
+      address = recordAddress
     }
```
will do

lucteo committed:
Revert "Simplify access to current `Function` in `Emitter`"

This reverts commit 6190c6b2918dcd80f4b095c9188672d45af84801.

lucteo committed:
Add `irType(of:)` in `Emitter` and replaced some of caching of `module[insertionFunction!]` to use this function.

We still kept the caching of `module[insertionFunction!]` when we used in inside iterations.

lucteo committed:
Improve predicate checking on `Optional`

lucteo committed:
Simplify construction of `_subfield_view`

lucteo committed:
Improve `_project_bundle`

lucteo committed:
Add missing preconditions.

lucteo commented:
Abandoning.
